[C#] Improve Documentation for Typed `Godot.Collections` Wrappers

Delsin-Yu opened the pull request:
Closes #102753 

This pull request includes updates to the documentation comments for the `Godot.Collections.Array<[MustBeVariant] T>` and `Godot.Collections.Dictionary<[MustBeVariant] TKey, [MustBeVariant] TValue>` wrapper types to improve clarity and accuracy.

Documentation improvements:

* [`modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs`](diffhunk://#diff-7b222db775bbaf1eb3d1102d64652196b35cef13053ca8b5a92cba18a1fd2bd2L1038-R1038): Updated the summary comment to specify that the array contains elements of type `<typeparamref name="T"/>`.
* [`modules/mono/glue/GodotSharp/GodotSharp/Core/Dictionary.cs`](diffhunk://#diff-6ad5c52c284ba0bb55706e339e12e6a0f0ece132a0bd9cb4f67203937077223cL479-R479): Updated the summary comment to specify that the dictionary contains elements of types `<typeparamref name="TKey"/>` and `<typeparamref name="TValue"/>`.

Delsin-Yu commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Dictionary.cs:
```diff
@@ -476,7 +476,7 @@ internal interface IGenericGodotDictionary
 
     /// <summary>
-    /// Typed wrapper around Godot's Dictionary class, a dictionary of Variant
+    /// Typed wrapper around Godot's Dictionary class, a dictionary of <typeparamref name="TKey"/> and <typeparamref name="TValue"/>
```
I don't think the line break style is correct here. However, we are skipping the line break check for C# source files, so neither `pre-commit` nor `CI` is triggering. What would be a better approach here?

raulsntos reviewed (commented):
I don't think this change is correct. The C# types are just converting the `Variant` elements to `T` every time an element is added/removed/accessed from the collection, the underlying type stored is still `Variant`.

To be clear about what the documentation was saying:

> Typed wrapper around Godot's Array class, an array of Variant typed elements allocated in the engine in C+++.

This is describing the native C++ `Array` type that the C# type is a wrapper of. Even when the array is typed, the underlying storage is still `Variant` (`TypedArray<int32_t>` stores the `int32_t` elements as `Variant`).

This is different from packed arrays that actually store the element they are specifically designed to hold. For example, a `PackedByteArray` stores the elements as `byte`. As a result `PackedByteArray` can take less memory than `PackedInt32Array`, but `TypedArray<uint8_t>` and `TypedArray<int32_t>` will both take the same memory.

Common C# collection types such as `List<T>` or `T[]` work more like packed arrays in this sense, so I can see where the confusion may be coming from, but this is also why I think it's important to mention explicitly in the documentation that the elements are stored as `Variant` and not `T`. And all of this also applies to dictionaries.

Perhaps there's a better way to describe how these types work, but it should still be clear that the underlying type stored is `Variant` even if conceptually it feels like they store `T` elements.

Delsin-Yu committed:
Improve documentation for C# typed collection wrappers

Delsin-Yu commented:
I have updated the documentation to clarify that the elements are only "annotated" rather than "typed"; I also added remark blocks to indicate that regardless of the type used, the elements maintain the same memory footprint.

paulloz reviewed (approved):
Looks good to me 👍

Repiteo commented:
Thanks!
